Source/Core/Core/IOS/USB/Host.cpp

```diff
@@ -0,0 +1,238 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+
+#ifdef __LIBUSB__
+#include <libusb.h>
+#endif
+
+#include "Common/Assert.h"
+#include "Common/ChunkFile.h"
+#include "Common/CommonTypes.h"
+#include "Common/Logging/Log.h"
+#include "Common/Thread.h"
+#include "Core/ConfigManager.h"
+#include "Core/Core.h"
+#include "Core/IOS/USB/Common.h"
+#include "Core/IOS/USB/Host.h"
+#include "Core/IOS/USB/LibusbDevice.h"
+
+namespace IOS
+{
+namespace HLE
+{
+namespace Device
+{
+USBHost::USBHost(u32 device_id, const std::string& device_name) : Device(device_id, device_name)
+{
+#ifdef __LIBUSB__
+  const int ret = libusb_init(&m_libusb_context);
+  _assert_msg_(WII_IPC_USB, ret == 0, "Failed to init libusb.");
+  libusb_set_debug(m_libusb_context, LIBUSB_LOG_LEVEL_WARNING);
+#endif
+}
+
+USBHost::~USBHost()
+{
+#ifdef __LIBUSB__
+  libusb_exit(m_libusb_context);
+#endif
+}
+
+ReturnCode USBHost::Open(const OpenRequest& request)
+{
+  // Force a device scan to complete, because some games (including Your Shape) only care
+  // about the initial device list (in the first GETDEVICECHANGE reply).
+  while (!UpdateDevices())
+  {
+  }
+  StartThreads();
+  return IPC_SUCCESS;
+}
+
+void USBHost::UpdateWantDeterminism(const bool new_want_determinism)
+{
+  if (new_want_determinism)
+    StopThreads();
+  else if (IsOpened())
+    StartThreads();
+}
+
+void USBHost::DoState(PointerWrap& p)
+{
+  if (IsOpened() && p.GetMode() == PointerWrap::MODE_READ)
+  {
+    // After a state has loaded, there may be insertion hooks for devices that were
+    // already plugged in, and which need to be triggered.
+    UpdateDevices(true);
+  }
+}
+
+bool USBHost::AddDevice(std::unique_ptr<USB::Device> device)
+{
+  std::lock_guard<std::mutex> lk(m_devices_mutex);
+  if (m_devices.count(device->GetId()) != 0)
```

Comment:
This should use `find()`, since it's more indicative of what you're trying to do (find if anything already exists).